lens-labels/src/Lens/Labels.hs

```diff
@@ -0,0 +1,114 @@
+{- | A lens library that integrates with OverloadedLabels.
+
+Unlike the `lens` package (and others), lenses are defined as a newtype
+instead of a type synonym, to avoid overlapping with other OverloadedLabel
+instances.  However, the `LensFn` and `runLens` functions allow converting
+between the two types; for example:
+
+> LensFn :: Control.Lens.LensLike f s t a b -> Lens.Labels.LensLike f s t a b
+> runLens :: Lens.Labels.LensLike f s t a b -> Control.Lens.LensLike f s t a b
+
+TODO: support more general optic types (e.g., prisms).
+-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ExplicitNamespaces #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Lens.Labels (
+    -- * Lenses
+    LensFn(..),
+    LensLike,
+    LensLike',
+    (&),
+    (Category..),
+    Lens,
+    -- * HasLens
+    HasLens(..),
+    Proxy#(..),
+    proxy#,
+    -- * Setters
+    ASetter,
+    (.~),
+    (%~),
+    set,
+    over,
+    -- * Getters
+    Const(..),
+    Getting,
+    (^.),
+    view,
+    ) where
+
+import qualified Control.Category as Category
+import GHC.Prim (Proxy#, proxy#)
+#if __GLASGOW_HASKELL__ >= 800
+import GHC.OverloadedLabels (IsLabel(..))
+#endif
+import GHC.TypeLits (Symbol)
+
+import Data.Function ((&))
+#if __GLASGOW_HASKELL__ >= 800
+import Data.Functor.Const(Const(..))
+#endif
+import Data.Functor.Identity(Identity(..))
+
+
+-- | A newtype for defining lenses.  Can be composed using
+-- `(Control.Category..)` (also exported from this module).
+newtype LensFn a b = LensFn {runLens :: a -> b}
+
+instance Category.Category LensFn where
+    id = LensFn id
+    LensFn f . LensFn g = LensFn (f . g)
+
+type LensLike f s t a b = LensFn (a -> f b) (s -> f t)
+type LensLike' f s a = LensLike f s s a a
+type Lens s t a b = forall f . Functor f => LensLike f s t a b
+
+-- | A type class for lens fields.
+class HasLens (x :: Symbol) f s t a b
+        | x s -> a, x t -> b, x s b -> t, x t a -> s where
+    lensOf :: Proxy# x -> (a -> f b) -> s -> f t
+
+#if __GLASGOW_HASKELL__ >= 800
+instance
+    (p ~ (a -> f b), HasLens x f s t a b)
+    => IsLabel x (LensFn p (s -> f t)) where
+    fromLabel p = LensFn $ lensOf p
+#endif
+
+type ASetter s t a b = LensLike Identity s t a b
```

Comment:
Would it be better to wrap the interface of an existing lens library with appropriate 'coerce' calls instead of implementing a subset of the functions?